packages/client/lib/miner/miner.ts

```diff
@@ -0,0 +1,206 @@
+   */
+  constructor(options: MinerOptions) {
+    this.config = options.config
+    this.synchronizer = options.synchronizer
+    this.running = false
+    this.assembling = false
+    this.period = 15000 // default: 15s period defined in ms
+  }
+
+  /**
+   * Convenience alias to return the latest block in the blockchain
+   */
+  private async latestBlock() {
+    return this.synchronizer.execution.vm.blockchain.getLatestBlock()
+  }
+
+  /**
+   * Sets the timeout for the next block assembly
+   */
+  private queueNextAssembly(timeout?: number) {
+    this._nextAssemblyTimeoutId && clearTimeout(this._nextAssemblyTimeoutId)
+    this._nextAssemblyTimeoutId = setTimeout(this.assembleBlock.bind(this), timeout ?? this.period)
+  }
+
+  /**
+   * Sets the next block assembly to latestBlock.timestamp + period
+   */
+  private async chainUpdated() {
+    const latestBlock = await this.latestBlock()
+    const target = latestBlock.header.timestamp.muln(1000).addn(this.period).sub(new BN(Date.now()))
+    const timeout = BN.max(new BN(0), target).toNumber()
+    this.config.logger.debug(
+      `Miner: Chain updated with block ${latestBlock.header.number.toNumber()}. Queuing next block assembly in ${Math.round(
+        timeout / 1000
+      )}s`
+    )
+    this.queueNextAssembly(timeout)
+  }
+
+  /**
+   * Start miner
+   */
+  start(): boolean {
+    if (!this.config.mine || this.running) {
+      return false
+    }
+    this.running = true
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      this.period = this.config.chainCommon.consensusConfig().period * 1000
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, this.chainUpdated.bind(this))
+    this.queueNextAssembly()
+    this.config.logger.info(`Miner started. Assembling next block in ${this.period / 1000}s`)
+    return true
+  }
+
+  /**
+   * Assembles a block from txs in the TxPool and adds it to the chain.
+   * If a new block is received while assembling it will abort.
+   */
+  async assembleBlock() {
+    if (this.assembling) {
+      return
+    }
+    this.assembling = true
+
+    // Abort if a new block is received while assembling this block
+    let interrupt = false
+    const setInterrupt = () => {
+      interrupt = true
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, setInterrupt)
+
+    const parentBlock = await this.latestBlock()
+    const { gasLimit } = parentBlock.header
+    const [signerAddress, signerPrivKey] = this.config.accounts[0]
+
+    // Use a copy of the vm to not modify the existing state.
+    // The state will be updated when the newly assembled block
+    // is inserted into the canonical chain.
+    const vmCopy = this.synchronizer.execution.vm.copy()
+
+    // If we are on the genesis block the canonical
+    // genesis state has not been initialiezd yet.
+    if (parentBlock.header.number.isZero()) {
+      await vmCopy.stateManager.generateCanonicalGenesis()
+    }
```

Comment:
Already mentioned this somewhere on the side line, I guess this is pretty dangerous to get into some totally weird and undefined behavior when the chain is actually "existing" but we assume a non-existing chain here because sync has not yet started or something.

Geth has a dedicated `init` command for this to ensure that a new chain is not started just-by-accident:

![grafik](https://user-images.githubusercontent.com/931137/132208363-6f70752f-c690-4fbc-a2ae-d51df27832e4.png)

Do we want to add something similar here, so maybe an option `--initGenesis` or something and just do the above if the option is provided and otherwise throw?

Reply:
great suggestion. In this case it should be safe because we're only doing it on the VM copy to assemble a first valid block after genesis to ensure a matching end stateRoot.

This is mostly an edge case of our own VM execution setup where when starting up the blockchain hasn't been initialized properly yet - on the insertion of the first block it would run [`generateCanonicalGenesis()`](https://github.com/ethereumjs/ethereumjs-monorepo/blob/c008e8eb76f520df83eb47c769e3a006bc24124f/packages/client/lib/sync/execution/vmexecution.ts#L100) - but we need this to happen before starting the BlockBuilder otherwise we will be working on an invalid state and conclude with an invalid stateRoot.

Alternatively, I might be able to change the line here to just run the iterator instead, will investigate!

Reply:
Ah, thanks for the clarification.

Yeah, I would find it more intuitive if this "progressing" is done by the iterator but - not sure - if it's not fitting our design it might not be worth it atm.

Reply:
Yeah I think this is an improvement we can make to the design of vmexecution, to better initialize the vm to a more "ready" state when just starting with the genesis block, rather than needing a sync to start. This is uncharted territory since before we could only sync and now we can create our own chain :) so it's understandable to need some adjustments to these newer parts.